pkg/kubelet/pod_workers.go

```diff
@@ -40,40 +40,98 @@ import (
+	// is kill or terminated.
 	MirrorPod *v1.Pod
-	// the type of update (create, update, sync, kill)
+	// RunningPod is a runtime pod that is no longer present in config. Required
+	// if Pod is nil, ignored if Pod is set.
+	RunningPod *kubecontainer.Pod
+	// The type of update (create, update, sync, kill).
 	UpdateType kubetypes.SyncPodType
-	// optional callback function when operation completes
-	// this callback is not guaranteed to be completed since a pod worker may
-	// drop update requests if it was fulfilling a previous request.  this is
-	// only guaranteed to be invoked in response to a kill pod request which is
-	// always delivered.
+	// OnCompleteFunc is an optional callback function for when operation
+	// completes. This method is always called, but another operation may
+	// have occurred in the meantime so callers must check the latest status.
 	OnCompleteFunc OnCompleteFunc
-	// if update type is kill, use the specified options to kill the pod.
+	// KillPodOptions is used to override the default termination behavior of the
+	// pod or to update the pod status after an operation is completed. Since a
+	// pod can be killed for multiple reasons, PodStatusFunc is invoked in order
+	// and later kills have an opportunity to override the status (i.e. a preemption
+	// may be later turned into an eviction).
 	KillPodOptions *KillPodOptions
 }
 
 // PodWorkers is an abstract interface for testability.
 type PodWorkers interface {
-	UpdatePod(options *UpdatePodOptions)
-	ForgetNonExistingPodWorkers(desiredPods map[types.UID]sets.Empty)
-	ForgetWorker(uid types.UID)
+	// UpdatePod notifies the pod worker of a change to a pod, which will then
+	// be processed in FIFO order by a goroutine per pod UID. The state of the
+	// pod will be passed to the syncPod method until either the pod is marked
+	// as deleted, it reaches a terminal phase (Succeeded/Failed), or the pod
+	// is evicted by the kubelet. Once that occurs the syncTerminatingPod method
+	// will be called until it exits successfully, and after that all further
+	// UpdatePod() calls will be ignored for that pod until it has been forgotten
+	// due to significant time passing. A pod that is terminated will never be
+	// restarted.
+	UpdatePod(options UpdatePodOptions)
+	// SyncKnownPods removes workers for pods that are not in the desiredPods set
+	// and have been terminated for a significant period of time. Once this method
+	// has been called once, the workers are assumed to be fully initialized and
+	// subsequent calls to ShouldPodContentBeRemoved on unknown pods will return
+	// true.
+	SyncKnownPods(desiredPods []*v1.Pod) map[types.UID]kubelettypes.SyncPodType
+	// CouldHaveRunningContainers returns true before the pod workers have synced,
+	// once the pod workers see the pod (syncPod could be called), and returns false
+	// after the pod has been terminated (running containers guaranteed stopped).
+	CouldHaveRunningContainers(uid types.UID) bool
```

Comment:
I started with that path, and over repeated reads of the code found it very difficult to understand the loops people were using in that context.  Everyone besides core kubelet loop (volume, gc, etc) should probably be describing in terms of setup and teardown, where setup is racing against the sync pod loop, and teardown is racing against the termination loops. So I wanted to define these in terms of windows:

```
|------------------------------------------| (lifetime of pod object)
     |-------------| (syncing)
|--------------------------------| (acceptable to create)
...--|                           |------------------------------------------------- .... (acceptable to teardown some components)
...--|                                     |------------------------------------------- .... (acceptable to teardown all components)
```

I didn't feel Terminating or Terminated captured those last two accurately, and in fact were confusing (because you could maybe think terminated lasted up until the pod was deleted).

Reply:
> Terminating(), Terminated() and Deleting()

It's unclear to me what would be the difference between Terminating and Deleting, @Random-Liu 

Agreed that it does seem like there's a lot of new conditions defined, possibly more than needed.

@smarterclayton perhaps we should be signaling which are intended for internal use in the kubelet loop vs other consumers by use of public interface vs. private module-specific stuff?

Reply:
Terminating means the pod is shutting down.  Deleting means the pod is being deleted.  Those are orthogonal.  A deleted pod can aggressively have content removed by eviction.  A terminating pod has no content being removed.  

* Evicted
* Deleted
* Terminal (succeeded / failed)
* Termination requested
* Termination started
* Terminated
* Gone (orphan)

These are all different states.  Should* hides those conditions into consistent behaviors.